File: ashgf/functions/extended.py
"""Extended test functions for optimization benchmarks.

This module provides 19 standalone functions, each accepting a numpy array
``x`` and returning a scalar ``float``.  They are converted from the old
class-based ``Function`` implementation, with ``self.`` references and
``self.power(x)`` calls replaced by direct vectorised NumPy operations.
"""

from __future__ import annotations

import numpy as np

__all__ = [
    "extended_baele",
    "extended_bd1",
    "extended_cliff",
    "extended_denschnb",
    "extended_denschnf",
    "extended_feudenstein_and_roth",
    "extended_hiebert",
    "extended_himmelblau",
    "extended_maratos",
    "extended_penalty",
    "extended_psc1",
    "extended_quadratic_exponential_ep1",
    "extended_rosenbrock",
    "extended_tridiagonal_1",
    "extended_tridiagonal_2",
    "extended_trigonometric",
    "extended_white_and_holst",
    "generalized_rosenbrock",
    "generalized_white_and_holst",
]


# ---------------------------------------------------------------------------
# 1. extended_feudenstein_and_roth
# ---------------------------------------------------------------------------


def extended_feudenstein_and_roth(x: np.ndarray) -> float:
    """Extended Freudenstein & Roth function.

    Parameters
    ----------
    x : np.ndarray
        Input vector of even length.

    Returns
    -------
    float
        Sum of squared terms over pairs ``(x_p, x_d)``.
    """
    x_p = x[:-1:2]
    x_d = x[1::2]

    inner = ((5.0 - x_d) * x_d - 2.0) * x_d
    term_1 = (-13.0 + x_p + inner) ** 2
    term_2 = (-29.0 + x_p + inner) ** 2

    return float(np.sum(term_1 + term_2))


# ---------------------------------------------------------------------------
# 2. extended_trigonometric
# ---------------------------------------------------------------------------


def extended_trigonometric(x: np.ndarray) -> float:
    """Extended Trigonometric function.

    Parameters
    ----------
    x : np.ndarray
        Input vector of arbitrary length ``n``.

    Returns
    -------
    float
        Sum of squared trigonometric terms.
    """
    n = len(x)
    cos_x = np.cos(x)

    term_1 = n - np.sum(cos_x)  # scalar
    term_2 = np.arange(1, n + 1) * (1.0 - cos_x)  # array
    term_3 = np.sin(x)  # array

    return float(np.sum((term_1 + term_2 + term_3) ** 2))


# ---------------------------------------------------------------------------
# 3. extended_rosenbrock
# ---------------------------------------------------------------------------


def extended_rosenbrock(x: np.ndarray) -> float:
    """Extended Rosenbrock function (pairwise variant).

    Parameters
    ----------
    x : np.ndarray
        Input vector of even length.

    Returns
    -------
    float
        Sum of Rosenbrock terms over consecutive pairs ``(x_p, x_d)``.
    """
    x_p = x[:-1:2]
    x_d = x[1::2]
    c = 100.0

    term_1 = c * (x_d - x_p**2) ** 2
    term_2 = (1.0 - x_p) ** 2

    return float(np.sum(term_1 + term_2))


# ---------------------------------------------------------------------------
# 4. generalized_rosenbrock
# ---------------------------------------------------------------------------


def generalized_rosenbrock(x: np.ndarray) -> float:
    """Generalized Rosenbrock function (chain variant).

    Parameters
    ----------
    x : np.ndarray
        Input vector of arbitrary length ``n >= 2``.

    Returns
    -------
    float
        Sum of Rosenbrock terms along the chain ``x[i] -> x[i+1]``.
    """
    c = 100.0

    term_1 = c * (x[1:] - x[:-1] ** 2) ** 2
    term_2 = (1.0 - x[:-1]) ** 2

    return float(np.sum(term_1 + term_2))


# ---------------------------------------------------------------------------
# 5. extended_white_and_holst
# ---------------------------------------------------------------------------


def extended_white_and_holst(x: np.ndarray) -> float:
    """Extended White & Holst function (pairwise variant).

    Parameters
    ----------
    x : np.ndarray
        Input vector of even length.

    Returns
    -------
    float
        Sum of White & Holst terms over pairs ``(x_p, x_d)``.
    """
    x_p = x[:-1:2]
    x_d = x[1::2]
    c = 100.0

    term_1 = c * (x_d - x_p**3) ** 2
    term_2 = (1.0 - x_p) ** 2

    return float(np.sum(term_1 + term_2))


# ---------------------------------------------------------------------------
# 6. extended_baele
# ---------------------------------------------------------------------------


def extended_baele(x: np.ndarray) -> float:
    """Extended Baele function.

    Parameters
    ----------
    x : np.ndarray
        Input vector of even length.

    Returns
    -------
    float
        Sum of three squared residual terms over pairs ``(x_p, x_d)``.
    """
    x_p = x[:-1:2]
    x_d = x[1::2]

    term_1 = (1.5 - x_p * (1.0 - x_d)) ** 2
    term_2 = (2.25 - x_p * (1.0 - x_d**2)) ** 2
    term_3 = (2.625 - x_p * (1.0 - x_d**3)) ** 2

    return float(np.sum(term_1 + term_2 + term_3))


# ---------------------------------------------------------------------------
# 7. extended_penalty
# ---------------------------------------------------------------------------


def extended_penalty(x: np.ndarray) -> float:
    """Extended Penalty function.

    Parameters
    ----------
    x : np.ndarray
        Input vector of arbitrary length.

    Returns
    -------
    float
        Penalty value combining element-wise and global constraint violations.
    """
    term_1 = (x[:-1] - 1.0) ** 2
    term_2 = (np.sum(x**2) - 0.25) ** 2

    return float(np.sum(term_1) + term_2)


# ---------------------------------------------------------------------------
# 8. extended_himmelblau
# ---------------------------------------------------------------------------


def extended_himmelblau(x: np.ndarray) -> float:
    """Extended Himmelblau function.

    Parameters
    ----------
    x : np.ndarray
        Input vector of even length.

    Returns
    -------
    float
        Sum of Himmelblau terms over pairs ``(x_p, x_d)``.
    """
    x_p = x[:-1:2]
    x_d = x[1::2]

    term_1 = (x_p**2 + x_d - 11.0) ** 2
    term_2 = (x_p + x_d**2 - 7.0) ** 2

    return float(np.sum(term_1 + term_2))


# ---------------------------------------------------------------------------
# 9. generalized_white_and_holst
# ---------------------------------------------------------------------------


def generalized_white_and_holst(x: np.ndarray) -> float:
    """Generalized White & Holst function (chain variant).

    Parameters
    ----------
    x : np.ndarray
        Input vector of arbitrary length ``n >= 2``.

    Returns
    -------
    float
        Sum of White & Holst terms along the chain ``x[i] -> x[i+1]``.
    """
    c = 100.0

    term_1 = c * (x[1:] - x[:-1] ** 3) ** 2
    term_2 = (1.0 - x[:-1]) ** 2

    return float(np.sum(term_1 + term_2))


# ---------------------------------------------------------------------------
# 10. extended_psc1
# ---------------------------------------------------------------------------


def extended_psc1(x: np.ndarray) -> float:
    """Extended PSC1 function.

    Parameters
    ----------
    x : np.ndarray
        Input vector of even length.

    Returns
    -------
    float
        Sum of polynomial and trigonometric terms over pairs ``(x_p, x_d)``.
    """
    x_p = x[:-1:2]
    x_d = x[1::2]

    term_1 = (x_p**2 + x_d**2 + x_p * x_d) ** 2
    term_2 = np.sin(x_p) ** 2
    term_3 = np.cos(x_d) ** 2

    return float(np.sum(term_1 + term_2 + term_3))


# ---------------------------------------------------------------------------
# 11. extended_bd1
# ---------------------------------------------------------------------------


def extended_bd1(x: np.ndarray) -> float:
    """Extended BD1 function.

    Parameters
    ----------
    x : np.ndarray
        Input vector of even length.

    Returns
    -------
    float
        Sum of polynomial and exponential residual terms over pairs.
    """
    x_p = x[:-1:2]
    x_d = x[1::2]

    term_1 = (x_p**2 + x_d - 2.0) ** 2
    term_2 = (np.exp(x_p - 1.0) - x_p) ** 2

    return float(np.sum(term_1 + term_2))


# ---------------------------------------------------------------------------
# 12. extended_maratos
# ---------------------------------------------------------------------------


def extended_maratos(x: np.ndarray) -> float:
    """Extended Maratos function.

    Parameters
    ----------
    x : np.ndarray
        Input vector of even length.

    Returns
    -------
    float
        Sum of linear and quadratic penalty terms over pairs.
    """
    x_p = x[:-1:2]
    x_d = x[1::2]
    c = 100.0

    term_1 = x_p
    term_2 = c * (x_p**2 + x_d**2 - 1.0) ** 2

    return float(np.sum(term_1 + term_2))


# ---------------------------------------------------------------------------
# 13. extended_cliff
# ---------------------------------------------------------------------------


def extended_cliff(x: np.ndarray) -> float:
    """Extended Cliff function.

    Parameters
    ----------
    x : np.ndarray
        Input vector of even length.

    Returns
    -------
    float
        Sum of quadratic, linear, and exponential cliff terms over pairs.
    """
    x_p = x[:-1:2]
    x_d = x[1::2]

    term_1 = ((x_p - 3.0) / 100.0) ** 2
    term_2 = x_p - x_d
    # Clip the exponent to prevent overflow in exp(20 * diff)
    cliff_exp = np.exp(np.clip(20.0 * (x_p - x_d), -100.0, 100.0))

    return float(np.sum(term_1 + term_2 + cliff_exp))


# ---------------------------------------------------------------------------
# 14. extended_hiebert
# ---------------------------------------------------------------------------


def extended_hiebert(x: np.ndarray) -> float:
    """Extended Hiebert function.

    Parameters
    ----------
    x : np.ndarray
        Input vector of even length.

    Returns
    -------
    float
        Sum of large-scale quadratic residual terms over pairs.
    """
    x_p = x[:-1:2]
    x_d = x[1::2]

    term_1 = (x_p - 10.0) ** 2
    term_2 = (x_p * x_d - 50000.0) ** 2

    return float(np.sum(term_1 + term_2))


# ---------------------------------------------------------------------------
# 15. extended_tridiagonal_1
# ---------------------------------------------------------------------------


def extended_tridiagonal_1(x: np.ndarray) -> float:
    """Extended Tridiagonal 1 function.

    Parameters
    ----------
    x : np.ndarray
        Input vector of even length.

    Returns
    -------
    float
        Sum of quadratic and quartic terms over pairs ``(x_p, x_d)``.
    """
    x_p = x[:-1:2]
    x_d = x[1::2]

    term_1 = (x_p + x_d - 3.0) ** 2
    term_2 = (x_p - x_d + 1.0) ** 4

    return float(np.sum(term_1 + term_2))


# ---------------------------------------------------------------------------
# 16. extended_tridiagonal_2
# ---------------------------------------------------------------------------


def extended_tridiagonal_2(x: np.ndarray) -> float:
    """Extended Tridiagonal 2 function (chain variant).

    Parameters
    ----------
    x : np.ndarray
        Input vector of arbitrary length ``n >= 2``.

    Returns
    -------
    float
        Sum of product-based quadratic terms along the chain.
    """
    c = 0.1

    term_1 = (x[1:] * x[:-1] - 1.0) ** 2
    term_2 = c * (x[:-1] + 1.0) ** 2

    return float(np.sum(term_1 + term_2))


# ---------------------------------------------------------------------------
# 17. extended_denschnb
# ---------------------------------------------------------------------------


def extended_denschnb(x: np.ndarray) -> float:
    """Extended Denschnb function.

    Parameters
    ----------
    x : np.ndarray
        Input vector of even length.

    Returns
    -------
    float
        Sum of quadratic and mixed terms over pairs ``(x_p, x_d)``.
    """
    x_p = x[:-1:2]
    x_d = x[1::2]

    term_1 = (x_p - 2.0) ** 2
    term_2 = term_1 * x_d**2
    term_3 = (x_d + 1.0) ** 2

    return float(np.sum(term_1 + term_2 + term_3))


# ---------------------------------------------------------------------------
# 18. extended_denschnf
# ---------------------------------------------------------------------------


def extended_denschnf(x: np.ndarray) -> float:
    """Extended Denschnf function.

    Parameters
    ----------
    x : np.ndarray
        Input vector of even length.

    Returns
    -------
    float
        Sum of squared quadratic combinations and a polynomial penalty over pairs.
    """
    x_p = x[:-1:2]
    x_d = x[1::2]

    term_1 = 2.0 * (x_p + x_d) ** 2
    term_2 = (x_p - x_d) ** 2 - 8.0
    term_3 = (5.0 * x_p**2 + (x_p - 3.0) ** 2 - 9.0) ** 2

    return float(np.sum((term_1 + term_2) ** 2 + term_3))


# ---------------------------------------------------------------------------
# 19. extended_quadratic_exponential_ep1
# ---------------------------------------------------------------------------


def extended_quadratic_exponential_ep1(x: np.ndarray) -> float:
    """Extended Quadratic Exponential EP1 function.

    Parameters
    ----------
    x : np.ndarray
        Input vector of even length.

    Returns
    -------
    float
        Sum of exponential, quadratic, and shifted quadratic terms over pairs.
    """
    x_p = x[:-1:2]
    x_d = x[1::2]

    diff = x_p - x_d

    term_1 = (np.exp(diff) - 5.0) ** 2
    term_2 = diff**2
    term_3 = (diff - 11.0) ** 2

    return float(np.sum(term_1 + term_2 * term_3))

File: src/functions/extended.rs
//! Extended optimisation test functions (pairwise / chain variants).

use ndarray::Array1;

// ---------------------------------------------------------------------------
// Helper
// ---------------------------------------------------------------------------

fn cached_arange(n: usize) -> Array1<f64> {
    Array1::from_iter((1..=n).map(|i| i as f64))
}

// ---------------------------------------------------------------------------
// 1. extended_feudenstein_and_roth
// ---------------------------------------------------------------------------

pub fn extended_feudenstein_and_roth(x: &Array1<f64>) -> f64 {
    let n = x.len();
    let xp = x.slice(ndarray::s![0..n-1;2]);
    let xd = x.slice(ndarray::s![1..n;2]);
    let mut total = 0.0;
    for i in 0..xp.len().min(xd.len()) {
        let a = xp[i];
        let b = xd[i];
        let inner = ((5.0 - b) * b - 2.0) * b;
        let term1 = (-13.0 + a + inner).powi(2);
        let term2 = (-29.0 + a + inner).powi(2);
        total += term1 + term2;
    }
    total
}

// ---------------------------------------------------------------------------
// 2. extended_trigonometric
// ---------------------------------------------------------------------------

pub fn extended_trigonometric(x: &Array1<f64>) -> f64 {
    let n = x.len();
    let cos_x = x.mapv(|v| v.cos());
    let indices = cached_arange(n);

    let term1 = n as f64 - cos_x.sum();
    let term2 = &indices * &(1.0 - &cos_x);
    let term3 = x.mapv(|v| v.sin());

    (&term2 + &term3 + term1).mapv(|v| v.powi(2)).sum()
}

// ---------------------------------------------------------------------------
// 3. extended_rosenbrock
// ---------------------------------------------------------------------------

pub fn extended_rosenbrock(x: &Array1<f64>) -> f64 {
    let n = x.len();
    let xp = x.slice(ndarray::s![0..n-1;2]);
    let xd = x.slice(ndarray::s![1..n;2]);
    let c = 100.0;
    let mut total = 0.0;
    for i in 0..xp.len().min(xd.len()) {
        let a = xp[i];
        let b = xd[i];
        total += c * (b - a.powi(2)).powi(2) + (1.0 - a).powi(2);
    }
    total
}

// ---------------------------------------------------------------------------
// 4. generalized_rosenbrock (chain)
// ---------------------------------------------------------------------------

pub fn generalized_rosenbrock(x: &Array1<f64>) -> f64 {
    let c = 100.0;
    let n = x.len();
    if n < 2 {
        return 0.0;
    }
    let mut total = 0.0;
    for i in 0..n - 1 {
        let xi = x[i];
        let xj = x[i + 1];
        total += c * (xj - xi.powi(2)).powi(2) + (1.0 - xi).powi(2);
    }
    total
}

// ---------------------------------------------------------------------------
// 5. extended_white_and_holst
// ---------------------------------------------------------------------------

pub fn extended_white_and_holst(x: &Array1<f64>) -> f64 {
    let n = x.len();
    let xp = x.slice(ndarray::s![0..n-1;2]);
    let xd = x.slice(ndarray::s![1..n;2]);
    let c = 100.0;
    let mut total = 0.0;
    for i in 0..xp.len().min(xd.len()) {
        let a = xp[i];
        let b = xd[i];
        total += c * (b - a.powi(3)).powi(2) + (1.0 - a).powi(2);
    }
    total
}

// ---------------------------------------------------------------------------
// 6. extended_baele
// ---------------------------------------------------------------------------

pub fn extended_baele(x: &Array1<f64>) -> f64 {
    let n = x.len();
    let xp = x.slice(ndarray::s![0..n-1;2]);
    let xd = x.slice(ndarray::s![1..n;2]);
    let mut total = 0.0;
    for i in 0..xp.len().min(xd.len()) {
        let a = xp[i];
        let b = xd[i];
        total += (1.5 - a * (1.0 - b)).powi(2)
            + (2.25 - a * (1.0 - b.powi(2))).powi(2)
            + (2.625 - a * (1.0 - b.powi(3))).powi(2);
    }
    total
}

// ---------------------------------------------------------------------------
// 7. extended_penalty
// ---------------------------------------------------------------------------

pub fn extended_penalty(x: &Array1<f64>) -> f64 {
    let n = x.len();
    let term1: f64 = x
        .slice(ndarray::s![..n - 1])
        .iter()
        .map(|&v| (v - 1.0).powi(2))
        .sum();
    let term2 = (x.mapv(|v| v.powi(2)).sum() - 0.25).powi(2);
    term1 + term2
}

// ---------------------------------------------------------------------------
// 8. extended_himmelblau
// ---------------------------------------------------------------------------

pub fn extended_himmelblau(x: &Array1<f64>) -> f64 {
    let n = x.len();
    let xp = x.slice(ndarray::s![0..n-1;2]);
    let xd = x.slice(ndarray::s![1..n;2]);
    let mut total = 0.0;
    for i in 0..xp.len().min(xd.len()) {
        let a = xp[i];
        let b = xd[i];
        total += (a.powi(2) + b - 11.0).powi(2) + (a + b.powi(2) - 7.0).powi(2);
    }
    total
}

// ---------------------------------------------------------------------------
// 9. generalized_white_and_holst (chain)
// ---------------------------------------------------------------------------

pub fn generalized_white_and_holst(x: &Array1<f64>) -> f64 {
    let c = 100.0;
    let n = x.len();
    if n < 2 {
        return 0.0;
    }
    let mut total = 0.0;
    for i in 0..n - 1 {
        let xi = x[i];
        let xj = x[i + 1];
        total += c * (xj - xi.powi(3)).powi(2) + (1.0 - xi).powi(2);
    }
    total
}

// ---------------------------------------------------------------------------
// 10. extended_psc1
// ---------------------------------------------------------------------------

pub fn extended_psc1(x: &Array1<f64>) -> f64 {
    let n = x.len();
    let xp = x.slice(ndarray::s![0..n-1;2]);
    let xd = x.slice(ndarray::s![1..n;2]);
    let mut total = 0.0;
    for i in 0..xp.len().min(xd.len()) {
        let a = xp[i];
        let b = xd[i];
        total += (a.powi(2) + b.powi(2) + a * b).powi(2) + a.sin().powi(2) + b.cos().powi(2);
    }
    total
}

// ---------------------------------------------------------------------------
// 11. extended_bd1
// ---------------------------------------------------------------------------

pub fn extended_bd1(x: &Array1<f64>) -> f64 {
    let n = x.len();
    let xp = x.slice(ndarray::s![0..n-1;2]);
    let xd = x.slice(ndarray::s![1..n;2]);
    let mut total = 0.0;
    for i in 0..xp.len().min(xd.len()) {
        let a = xp[i];
        let b = xd[i];
        total += (a.powi(2) + b - 2.0).powi(2) + ((a - 1.0).exp() - a).powi(2);
    }
    total
}

// ---------------------------------------------------------------------------
// 12. extended_maratos
// ---------------------------------------------------------------------------

pub fn extended_maratos(x: &Array1<f64>) -> f64 {
    let n = x.len();
    let xp = x.slice(ndarray::s![0..n-1;2]);
    let xd = x.slice(ndarray::s![1..n;2]);
    let c = 100.0;
    let mut total = 0.0;
    for i in 0..xp.len().min(xd.len()) {
        let a = xp[i];
        let b = xd[i];
        total += a + c * (a.powi(2) + b.powi(2) - 1.0).powi(2);
    }
    total
}

// ---------------------------------------------------------------------------
// 13. extended_cliff
// ---------------------------------------------------------------------------

pub fn extended_cliff(x: &Array1<f64>) -> f64 {
    let n = x.len();
    let xp = x.slice(ndarray::s![0..n-1;2]);
    let xd = x.slice(ndarray::s![1..n;2]);
    let mut total = 0.0;
    for i in 0..xp.len().min(xd.len()) {
        let a = xp[i];
        let b = xd[i];
        let cliff_exp = (20.0 * (a - b)).clamp(-100.0, 100.0).exp();
        total += ((a - 3.0) / 100.0).powi(2) + (a - b) + cliff_exp;
    }
    total
}

// ---------------------------------------------------------------------------
// 14. extended_hiebert
// ---------------------------------------------------------------------------

pub fn extended_hiebert(x: &Array1<f64>) -> f64 {
    let n = x.len();
    let xp = x.slice(ndarray::s![0..n-1;2]);
    let xd = x.slice(ndarray::s![1..n;2]);
    let mut total = 0.0;
    for i in 0..xp.len().min(xd.len()) {
        let a = xp[i];
        let b = xd[i];
        total += (a - 10.0).powi(2) + (a * b - 50000.0).powi(2);
    }
    total
}

// ---------------------------------------------------------------------------
// 15. extended_tridiagonal_1
// ---------------------------------------------------------------------------

pub fn extended_tridiagonal_1(x: &Array1<f64>) -> f64 {
    let n = x.len();
    let xp = x.slice(ndarray::s![0..n-1;2]);
    let xd = x.slice(ndarray::s![1..n;2]);
    let mut total = 0.0;
    for i in 0..xp.len().min(xd.len()) {
        let a = xp[i];
        let b = xd[i];
        total += (a + b - 3.0).powi(2) + (a - b + 1.0).powi(4);
    }
    total
}

// ---------------------------------------------------------------------------
// 16. extended_tridiagonal_2 (chain)
// ---------------------------------------------------------------------------

pub fn extended_tridiagonal_2(x: &Array1<f64>) -> f64 {
    let c = 0.1;
    let n = x.len();
    if n < 2 {
        return 0.0;
    }
    let mut total = 0.0;
    for i in 0..n - 1 {
        let xi = x[i];
        let xj = x[i + 1];
        total += (xj * xi - 1.0).powi(2) + c * (xi + 1.0).powi(2);
    }
    total
}

// ---------------------------------------------------------------------------
// 17. extended_denschnb
// ---------------------------------------------------------------------------

pub fn extended_denschnb(x: &Array1<f64>) -> f64 {
    let n = x.len();
    let xp = x.slice(ndarray::s![0..n-1;2]);
    let xd = x.slice(ndarray::s![1..n;2]);
    let mut total = 0.0;
    for i in 0..xp.len().min(xd.len()) {
        let a = xp[i];
        let b = xd[i];
        let term1 = (a - 2.0).powi(2);
        let term2 = term1 * b.powi(2);
        let term3 = (b + 1.0).powi(2);
        total += term1 + term2 + term3;
    }
    total
}

// ---------------------------------------------------------------------------
// 18. extended_denschnf
// ---------------------------------------------------------------------------

pub fn extended_denschnf(x: &Array1<f64>) -> f64 {
    let n = x.len();
    let xp = x.slice(ndarray::s![0..n-1;2]);
    let xd = x.slice(ndarray::s![1..n;2]);
    let mut total = 0.0;
    for i in 0..xp.len().min(xd.len()) {
        let a = xp[i];
        let b = xd[i];
        let term1 = 2.0 * (a + b).powi(2);
        let term2 = (a - b).powi(2) - 8.0;
        let term3 = (5.0 * a.powi(2) + (a - 3.0).powi(2) - 9.0).powi(2);
        total += (term1 + term2).powi(2) + term3;
    }
    total
}

// ---------------------------------------------------------------------------
// 19. extended_quadratic_exponential_ep1
// ---------------------------------------------------------------------------

pub fn extended_quadratic_exponential_ep1(x: &Array1<f64>) -> f64 {
    let n = x.len();
    let xp = x.slice(ndarray::s![0..n-1;2]);
    let xd = x.slice(ndarray::s![1..n;2]);
    let mut total = 0.0;
    for i in 0..xp.len().min(xd.len()) {
        let a = xp[i];
        let b = xd[i];
        let diff = a - b;
        total += (diff.exp() - 5.0).powi(2) + diff.powi(2) * (diff - 11.0).powi(2);
    }
    total
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rosenbrock_global_min() {
        // Extended Rosenbrock minimum at xp=1, xd=1 for all pairs
        let x = Array1::from_vec(vec![1.0, 1.0, 1.0, 1.0]);
        assert!((extended_rosenbrock(&x) - 0.0).abs() < 1e-10);
    }

    #[test]
    fn penalty_is_finite() {
        let x = Array1::zeros(10);
        assert!(extended_penalty(&x).is_finite());
    }

    #[test]
    fn himmelblau_known_min() {
        // One of the four minima: (3.0, 2.0)
        let x = Array1::from_vec(vec![3.0, 2.0]);
        assert!((extended_himmelblau(&x) - 0.0).abs() < 1e-8);
    }
}
